File: src/assembler/codex-config-assembler.ts
/**
 * CodexConfigAssembler — generates .codex/config.toml from Nunjucks template.
 *
 * Operates in 2 phases:
 * 1. Derivation — computes model, approval_policy, sandbox_mode, maps MCP servers
 * 2. Rendering — passes derived values to config.toml.njk and writes output
 *
 * @module
 */
import * as fs from "node:fs";
import * as path from "node:path";
import type { ProjectConfig } from "../models.js";
import type { TemplateEngine } from "../template-engine.js";
import { buildDefaultContext } from "../template-engine.js";
import type { AssembleResult } from "./rules-assembler.js";
import {
  DEFAULT_MODEL,
  SANDBOX_WORKSPACE_WRITE,
  detectHooks,
  deriveApprovalPolicy,
  mapMcpServers,
  isValidTomlBareKey,
} from "./codex-shared.js";

const TEMPLATE_PATH = "codex-templates/config.toml.njk";

/**
 * Build the template context for config.toml rendering.
 *
 * @param config - Project configuration.
 * @param hasHooks - Whether hooks were detected.
 * @returns Template context with all fields needed by config.toml.njk.
 */
export function buildConfigContext(
  config: ProjectConfig,
  hasHooks: boolean,
): Record<string, unknown> {
  const mcpServers = mapMcpServers(config);
  return {
    ...buildDefaultContext(config),
    model: DEFAULT_MODEL,
    approval_policy: deriveApprovalPolicy(hasHooks),
    sandbox_mode: SANDBOX_WORKSPACE_WRITE,
    mcp_servers: mcpServers,
    has_mcp: mcpServers.length > 0,
  };
}

/** Generates .codex/config.toml from Nunjucks template. */
export class CodexConfigAssembler {
  /** Generate .codex/config.toml by deriving config values and rendering template. */
  assemble(
    config: ProjectConfig,
    outputDir: string,
    _resourcesDir: string,
    engine: TemplateEngine,
  ): AssembleResult {
    const warnings: string[] = [];
    const hooksDir = path.join(
      path.dirname(outputDir), ".claude", "hooks",
    );
    const hasHooks = detectHooks(hooksDir);
    for (const server of config.mcp.servers) {
      if (!isValidTomlBareKey(server.id)) {
        warnings.push(
          `MCP server id "${server.id}" contains invalid TOML characters`,
        );
      }
    }
    const context = buildConfigContext(config, hasHooks);
    let rendered: string;
    try {
      rendered = engine.renderTemplate(TEMPLATE_PATH, context);
    } catch (error: unknown) {
      const message = error instanceof Error
        ? error.message
        : String(error);
      if (message.includes("not found") || message.includes("ENOENT")) {
        warnings.push(`Template not found: ${TEMPLATE_PATH}`);
        return { files: [], warnings };
      }
      throw error;
    }
    fs.mkdirSync(outputDir, { recursive: true });
    const dest = path.join(outputDir, "config.toml");
    fs.writeFileSync(dest, rendered, "utf-8");
    return { files: [dest], warnings };
  }
}

File: java/src/main/java/dev/iadev/assembler/CodexConfigAssembler.java
package dev.iadev.assembler;

import dev.iadev.config.ContextBuilder;
import dev.iadev.domain.model.McpServerConfig;
import dev.iadev.domain.model.ProjectConfig;
import dev.iadev.template.TemplateEngine;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Generates {@code .codex/config.toml} from Pebble template.
 *
 * <p>Operates in 2 phases:
 * <ol>
 *   <li>Derivation — computes model, approval_policy,
 *       sandbox_mode, maps MCP servers</li>
 *   <li>Rendering — passes derived values to
 *       {@code config.toml.njk} and writes output</li>
 * </ol>
 *
 * <p>This is the eighteenth assembler in the pipeline
 * (position 18 of 25 per RULE-005). Its target is
 * {@link AssemblerTarget#CODEX}.</p>
 *
 * @see Assembler
 * @see CodexShared
 */
public final class CodexConfigAssembler
        implements Assembler {

    private static final String TEMPLATE_PATH =
            "codex-templates/config.toml.njk";

    /**
     * {@inheritDoc}
     *
     * <p>Generates {@code config.toml} by deriving config
     * values and rendering the template.</p>
     */
    @Override
    public List<String> assemble(
            ProjectConfig config,
            TemplateEngine engine,
            Path outputDir) {
        return assembleWithResult(
                config, engine, outputDir).files();
    }

    /** {@inheritDoc} */
    @Override
    public AssemblerResult assembleWithResult(
            ProjectConfig config,
            TemplateEngine engine,
            Path outputDir) {
        Path hooksDir = outputDir.getParent()
                .resolve(".claude").resolve("hooks");
        Path agentsDir = outputDir.getParent()
                .resolve(".claude").resolve("agents");
        HookPresence hookPresence = HookPresence.of(
                CodexShared.detectHooks(hooksDir));
        List<AgentInfo> agents = CodexScanner.scanAgents(
                agentsDir);

        List<String> warnings =
                collectTomlKeyWarnings(config);

        Map<String, Object> context =
                buildConfigContext(
                        config, hookPresence, agents);

        String rendered = engine.render(
                TEMPLATE_PATH, context);

        CopyHelpers.ensureDirectory(outputDir);
        Path dest = outputDir.resolve("config.toml");
        CopyHelpers.writeFile(dest, rendered);

        return AssemblerResult.of(
                List.of(dest.toString()), warnings);
    }

    /**
     * Validates MCP server IDs for TOML bare key
     * compliance.
     *
     * @param config the project configuration
     * @return list of warning messages for invalid IDs
     */
    static List<String> collectTomlKeyWarnings(
            ProjectConfig config) {
        List<String> warnings = new ArrayList<>();
        for (McpServerConfig server
                : config.mcp().servers()) {
            if (!CodexShared.isValidTomlBareKey(
                    server.id())) {
                warnings.add(
                        ("MCP server id \"%s\" contains"
                                + " invalid TOML characters")
                                .formatted(server.id()));
            }
        }
        return warnings;
    }

    /**
     * Builds the template context for {@code config.toml}
     * rendering.
     *
     * @param config       the project configuration
     * @param hookPresence whether hooks were detected
     * @param agentsList   scanned Claude agents
     * @return the template context map
     */
    static Map<String, Object> buildConfigContext(
            ProjectConfig config,
            HookPresence hookPresence,
            List<AgentInfo> agentsList) {
        List<Map<String, Object>> mcpServers =
                CodexShared.mapMcpServers(config);
        List<Map<String, Object>> agents =
                mapAgents(agentsList);
        Map<String, Object> ctx = new LinkedHashMap<>(
                ContextBuilder.buildContext(config));
        ctx.put("model", CodexShared.DEFAULT_MODEL);
        ctx.put("approval_policy",
                CodexShared.deriveApprovalPolicy(
                        hookPresence));
        ctx.put("sandbox_mode",
                CodexShared.SANDBOX_WORKSPACE_WRITE);
        ctx.put("mcp_servers", mcpServers);
        ctx.put("has_mcp", !mcpServers.isEmpty());
        ctx.put("agents_list", agents);
        return ctx;
    }

    private static List<Map<String, Object>> mapAgents(
            List<AgentInfo> scanned) {
        List<Map<String, Object>> result = new ArrayList<>();
        for (AgentInfo agent : scanned) {
            Map<String, Object> row = new LinkedHashMap<>();
            row.put("name", CodexShared.sanitizeTomlBareKey(
                    agent.name()));
            row.put("description", CodexShared.escapeTomlValue(
                    agent.description()));
            result.add(row);
        }
        return result;
    }

}
